command/meta.go

```diff
@@ -212,11 +226,96 @@ func (m *ApiMeta) initializeX509Config(client http.Client, clientCA []byte, cert
 	client.Transport.(*http.Transport).TLSClientConfig.PreferServerCipherSuites = true
 	client.Transport.(*http.Transport).TLSClientConfig.Certificates = []tls.Certificate{cert}
-	client.Transport.(*http.Transport).TLSClientConfig.InsecureSkipVerify = true // TODO(jacobkiefer): Add a flag this.
+	client.Transport.(*http.Transport).TLSClientConfig.InsecureSkipVerify = true // TODO(jacobkiefer): Add a flag for this.
 
 	return &client
 }
 
+func (m *ApiMeta) Authenticate() error {
+	auth := m.Config.Auth
+	if auth != nil && auth.Enabled && auth.OAuth2 != nil {
+		OAuth2 := auth.OAuth2
+		if !OAuth2.IsValid() {
+			return errors.New("Incorrect OAuth2 auth configuration.")
+		}
+
+		usr, err := user.Current()
+		if err != nil {
+			m.Ui.Error("Could not read current user from environment, failing.")
+			return err
+		}
+		credentialsLocation := filepath.Join(usr.HomeDir, ".spin", "credentials")
+		config := &oauth2.Config{
+			ClientID:     OAuth2.ClientId,
+			ClientSecret: OAuth2.ClientSecret,
+			RedirectURL:  "http://localhost:8085",
+			Scopes:       OAuth2.Scopes,
+			Endpoint: oauth2.Endpoint{
+				AuthURL:  OAuth2.AuthUrl,
+				TokenURL: OAuth2.TokenUrl,
+			},
+		}
+		var newToken *oauth2.Token
+
+		if _, err := os.Stat(credentialsLocation); err == nil {
+			// Look up cached credentials to save oauth2 roundtrip.
+			var token oauth2.Token
+			tokenBytes, err := ioutil.ReadFile(credentialsLocation)
+			if err != nil {
+				m.Ui.Error(fmt.Sprintf("Could not read credentials file at %s", credentialsLocation))
+				return err
+			}
+			err = json.Unmarshal(tokenBytes, &token)
+			if err != nil {
+				m.Ui.Error(fmt.Sprintf("Could not parse OAuth2 token: %s", string(tokenBytes)))
+				return err
+			}
+			fmt.Printf("Read oauth2 token from cached file: %v\n", token)
+
+			tokenSource := config.TokenSource(oauth2.NoContext, &token)
+			if err != nil {
+				m.Ui.Error(fmt.Sprintf("Could not create token source from token: %v", token))
+				return err
+			}
+			newToken, err = tokenSource.Token()
+			if err != nil {
+				m.Ui.Error(fmt.Sprintf("Could not refresh token from source: %v", tokenSource))
+				return err
+			}
+		} else {
+			// Do roundtrip.
+			http.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				code := r.FormValue("code")
+				fmt.Fprintln(w, code)
+			}))
+			go http.ListenAndServe(":8085", nil)
+			// Note: leaving server connection open for scope of request, will be reaped on exit.
+
+			authURL := config.AuthCodeURL("state-token", oauth2.AccessTypeOffline, oauth2.ApprovalForce)
+			m.Ui.Output(fmt.Sprintf("Navigate to %s and authenticate", authURL))
+			code := m.Prompt()
+
+			newToken, err = config.Exchange(oauth2.NoContext, code)
+			if err != nil {
+				return err
+			}
+		}
+
+		m.Ui.Output("Caching oauth2 token.")
+		buf, _ := json.Marshal(&newToken)
+		ioutil.WriteFile(credentialsLocation, buf, 0777) // TODO: Which file mode?
```

Comment:
Can you lookup the filemode above? I probably would recommend against merging this until we have that figured out, since this makes the creds globally readable which isn't great.

Reply:
Will experiment for the correct filemode.